packages/opentelemetry-plugin-http/src/http.ts

```diff
@@ -0,0 +1,413 @@
+      };
+
+      const spanContext = propagation.extract(Format.HTTP, headers);
+      if (spanContext && isValid(spanContext)) {
+        spanOptions.parent = spanContext;
+      }
+
+      const span = plugin._startHttpSpan(`${method} ${pathname}`, spanOptions);
+
+      return plugin._tracer.withSpan(span, () => {
+        plugin._tracer.bind(request);
+        plugin._tracer.bind(response);
+
+        // Wraps end (inspired by:
+        // https://github.com/GoogleCloudPlatform/cloud-trace-nodejs/blob/master/src/plugins/plugin-connect.ts#L75)
+        const originalEnd = response.end;
+        response.end = function(
+          this: ServerResponse,
+          ...args: ResponseEndArgs
+        ) {
+          response.end = originalEnd;
+          // Cannot pass args of type ResponseEndArgs,
+          // tslint complains "Expected 1-2 arguments, but got 1 or more.", it does not make sense to me
+          // tslint:disable-next-line:no-any
+          const returned = response.end.apply(this, arguments as any);
+          const requestUrl = request.url ? url.parse(request.url) : null;
+          const hostname = headers.host
+            ? headers.host.replace(/^(.*)(\:[0-9]{1,5})/, '$1')
+            : 'localhost';
+          const userAgent = (headers['user-agent'] ||
+            headers['User-Agent']) as string;
+
+          const attributes: Attributes = {
+            [AttributeNames.HTTP_URL]: Utils.getUrlFromIncomingRequest(
+              requestUrl,
+              headers
+            ),
+            [AttributeNames.HTTP_HOSTNAME]: hostname,
+            [AttributeNames.HTTP_METHOD]: method,
+            [AttributeNames.HTTP_STATUS_CODE]: response.statusCode,
+            [AttributeNames.HTTP_STATUS_TEXT]: response.statusMessage,
+          };
+
+          if (requestUrl) {
+            attributes[AttributeNames.HTTP_PATH] = requestUrl.path || '/';
+            attributes[AttributeNames.HTTP_ROUTE] = requestUrl.pathname || '/';
+          }
+
+          if (userAgent) {
+            attributes[AttributeNames.HTTP_USER_AGENT] = userAgent;
+          }
+
+          span
+            .setAttributes(attributes)
+            .setStatus(Utils.parseResponseStatus(response.statusCode));
+
+          if (plugin.options.applyCustomAttributesOnSpan) {
+            plugin.options.applyCustomAttributesOnSpan(span, request, response);
+          }
+
+          span.end();
+          return returned;
+        };
+        return original.apply(this, [event, ...args]);
+      });
+    };
+  }
+
+  private _outgoingRequestFunction(
+    original: Func<ClientRequest>
+  ): Func<ClientRequest> {
+    const plugin = this;
+    return function outgoingRequest(
+      this: {},
+      options: RequestOptions | string,
+      ...args: unknown[]
+    ): ClientRequest {
+      if (!options) {
+        return original.apply(this, [options, ...args]);
+      }
+
+      const { origin, pathname, method, optionsParsed } = Utils.getRequestInfo(
```

Comment:
Recent NodeJs versions allow to pass `url` as `string` and a `options` object as second argument which seems to be not covered here.
see https://nodejs.org/dist/latest-v12.x/docs/api/http.html#http_http_request_url_options_callback